### PR TITLE
DNN-17683:  SI: DDR Menu: Exclude Nodes

### PR DESCRIPTION
**Pull request by Mohtshum**
- Simplify the matching logic for filtering nodes
- Add new method to remove the nested or child node by recursion

### CODE REVIEW

File: DNN Platform/Modules/DDRMenu/MenuBase.cs [hunk 1]
```diff
@@ -197,22 +197,15 @@ private void FilterNodes(string nodeString, bool exclude)
...
-							}));
+					var nodeText2 = nodeText;                    
+                    filteredNodes.Add(RootNode.FindByNameOrId(nodeText2));
 				}
 			}
 
             // if filtered for foksonomy tags, use flat tree to get all related pages in nodeselection
 		    if (flattenedNodes.HasChildren())
 		        RootNode = flattenedNodes;
-			RootNode.Children.RemoveAll(n => filteredNodes.Contains(n) == exclude);
```
Comment (tpluscode) [inline DNN Platform/Modules/DDRMenu/MenuBase.cs]: I see that the `exclude` parameter is not longer used in this method but it is called with both `true` and `false` in `MenuBase#PreRender`. If you're certain that your changes are correct for all cases please refactor the calling code too.

File: DNN Platform/Modules/DDRMenu/MenuBase.cs [hunk 2]
```diff
@@ -197,22 +197,15 @@ private void FilterNodes(string nodeString, bool exclude)
...
-					var nodeText2 = nodeText;
-					filteredNodes.AddRange(
-						RootNode.Children.FindAll(
-							n =>
-							{
-								var nodeName = n.Text.ToLowerInvariant();
-								var nodeId = n.TabId.ToString();
-								return (nodeText2 == nodeName || nodeText2 == nodeId);
-							}));
+					var nodeText2 = nodeText;
```
Comment (tpluscode) [inline DNN Platform/Modules/DDRMenu/MenuBase.cs]: This variable can be inlined.

File: DNN Platform/Modules/DDRMenu/MenuBase.cs (hunk 1 above)
Comment (Mohtshum) [reply DNN Platform/Modules/DDRMenu/MenuBase.cs]: Good catch that was surely missed out, thnx!

File: DNN Platform/Modules/DDRMenu/MenuBase.cs [hunk 3]
```diff
@@ -197,22 +197,21 @@ private void FilterNodes(string nodeString, bool exclude)
...
-								var nodeName = n.Text.ToLowerInvariant();
-								var nodeId = n.TabId.ToString();
-								return (nodeText2 == nodeName || nodeText2 == nodeId);
-							}));
+                    filteredNodes.Add(RootNode.FindByNameOrId(nodeText));
 				}
 			}
 
             // if filtered for foksonomy tags, use flat tree to get all related pages in nodeselection
 		    if (flattenedNodes.HasChildren())
```
Comment (tpluscode) [inline DNN Platform/Modules/DDRMenu/MenuBase.cs]: I know it's unrelated but please add braces here

File: DNN Platform/Modules/DDRMenu/MenuBase.cs [hunk 4]
```diff
@@ -197,22 +197,21 @@ private void FilterNodes(string nodeString, bool exclude)
...
 		        RootNode = flattenedNodes;
-			RootNode.Children.RemoveAll(n => filteredNodes.Contains(n) == exclude);
+            if (exclude)
+            {
+                RootNode.RemoveAll(filteredNodes);
```
Comment (tpluscode) [inline DNN Platform/Modules/DDRMenu/MenuBase.cs]: I think this condition is reversed. And it is suspicious that this call doesn't actually take the `exclude`  as parameter to pass down to it's children.

Comment (Mohtshum) [reply DNN Platform/Modules/DDRMenu/MenuBase.cs]: Actually in older code deletion was not recursive due to which issue was comming for true case i.e for EXCLUDE=true (when EXCLUDE is set) and when false things are working fine for other case

**Pull request merged**